KubernetesPodOperator doc (#175)
astronomer/docs: docs/kubernetespodoperator.md

---
sidebar_label: 'KubernetesPodOperator'
title: 'Run the KubernetesPodOperator on Astronomer Cloud'
id: kubernetespodoperator
---

## Overview

> The KubernetesPodOperator can run a task in an individual Kubernetes Pod. When the operator finishes running a task, it automatically terminates the Pod that the task was running in. The KubernetesPodOperator also enables task-level resource configuration, making it optimal for tasks that have custom Python dependencies.

Reviewer: Maybe also add a sentence here about how Astronomer Cloud is Kube based so you can do this without managing infra

> The KubernetesPodOperator can run a task in an individual Kubernetes Pod. When the operator finishes running a task, it automatically terminates the Pod that the task was running in. The KubernetesPodOperator also enables task-level resource configuration, making it optimal for tasks that have custom Python dependencies.

Author: Updated some sentences here and added a new bullet point for the benefit you commented

This guide provides steps for configuring and running the KubernetesPodOperator on DAGs deployed to Astronomer Cloud.

:::info

> This setup applies only to Astronomer projects running on Deployments in Astronomer Cloud.

Author: Gonna take this out for now until we merged the local docs in a separate PR

:::

## Prerequisites

To use the KubernetesPodOperator, you need:

- An [Astronomer Cloud project](create-project).

> - A [Deployment hosting your project](deploy-code).

Author: Had this when it was still assumed that users had to access their namespace from a Deployment information screen. Now that that's only an environment variable, this definitely makes sense.

- [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl).

> ## Using the KubernetesPodOperator

Reviewer: ```suggestion
## Import Statement
```

Author: Don't think it makes sense to have the import statement by itself, as you can't do much with that information alone.  imo anyone looking for the import statement will also look for how to configure the KPO

To start using the KubernetesPodOperator in a DAG, add the following import statements to your DAG file:

```python
from airflow.contrib.operators.kubernetes_pod_operator import kubernetes_pod_operator

# Pulls environment information from Astronomer Cloud
from airflow import configuration as conf
```

> You can now instantiate the KubernetesPodOperator:

Reviewer: ```suggestion
Then, instantiate the KubernetesPodOperator:
```

> You can now instantiate the KubernetesPodOperator:

Reviewer: Thoughts on adding an H2 here called `## Basic Configuration`, where we discuss the minimum configs required? AKA:

- Name
- Namespace
- Image
- task_id

Maybe an H3 under here that explains how the KPO actually runs on Cloud

And then another H2 after that for , `## Advanced Configuration`, and then H3's under that for:

- Resource Specifications 
- Pull Docker Image from a Private Registry
- Other use cases/configs

Author: I like changing the title of our core topic to "Configure the KubernetesPodOperator" and having another h2 later called "Advanced configuration". Would probably keep the private registry instructions separate from both of these tho, since that has more to do with administering Astronomer than running the KPO itself

```python
from airflow.contrib.operators.kubernetes_pod_operator import kubernetes_pod_operator

# Pulls environment information from Astronomer Cloud
from airflow import configuration as conf
...

namespace = conf.get('kubernetes', 'NAMESPACE')

> k = kubernetes_pod_operator.KubernetesPodOperator(
>     namespace=namespace,
>     image="ubuntu:16.04",
>     cmds=["bash", "-cx"],
>     arguments=["echo", "10", "echo pwd"],
>     labels={"foo": "bar"},
>     name="airflow-test-pod",
>     is_delete_operator_pod=True,
>     in_cluster=True,
>     task_id="task-two",
>     get_logs=True)

Reviewer: ```suggestion
k = kubernetes_pod_operator.KubernetesPodOperator(
    namespace=namespace, # Sets your Astronomer Deployment's namespace
    image="<your-docker-image>", # Specify your Docker image here
    cmds=["bash", "-cx"],
    arguments=["echo", "10", "echo pwd"],
    labels={"<pod-Label>": "<label-name>"}, # Labels that apply to the Kubernetes Pod
    name="<pod-name>", # The name of the Kubernetes Pod
    is_delete_operator_pod=True, # Instructs the pod to terminate after the task has been deleted
    in_cluster=True, # Instructs the pod to execute within the Kubernetes environment of your Astronomer Cluster
    task_id="<task-name>", # The name of your task
    get_logs=True) # If true, logs stdout output of the container
```

```

For each instantiation of the KubernetesPodOperator, you must specify the following values:

- `namespace = conf.get('kubernetes', 'NAMESPACE')`: Astronomer Cloud Deployments run on their own Kubernetes namespaces within a Cluster. The KubernetesPodOperator needs to know information about this namespace. Because the namespace variable is injected into your Deployment's `airflow.cfg` file, you can programmatically import this namespace information via environment variables using the statement `namespace = conf.get('kubernetes', 'NAMESPACE')`.
- `image`: This is the image that the operator will use to run its defined task, commands, and arguments. By default, this value should be a publicly available image tag on Dockerhub. If you need to pull an image from a private registry, read [Pull Images from a Private Registry](kubernetespodoperator#pull-images-from-a-private-registry).
- `in_cluster=True`: When this value is set, your task will look inside your Cluster for a Kubernetes configuration. This ensures that the workers running in the Pod are have the correct privileges within the Cluster.

> - `is_delete_operator_pod=True`: This setting ensures that the Pods of completed tasks are deleted, which lowers resource usage on your Cluster.

Reviewer: Thoughts on making this a table, where we set:

- Parameter
- Default
- Description

Think it's redundant with the `# comments` above?

Author: I think there's too much information here to put cleanly in a table, we should either:
- Have all info as inline comments in the code block and list just the minimum values (no descriptions) in a bulleted list
- Keep the code block clean and list all of the info about individual values in a bulleted list (this is how it currently is)

I'm cool with both!

---
sidebar_label: 'KubernetesPodOperator'
title: "Run Airflow's KubernetesPodOperator on Astronomer Cloud"
id: kubernetespodoperator
---

## Overview

One of Airflow's most dynamic operators is the [KubernetesPodOperator](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html), which can run a task in an individual Kubernetes pod. Similar to the Kubernetes Executor, this operator talks to the Kubernetes API to dynamically launch a pod for each task that needs to run and terminates each pod once the task is completed. This results in an isolated, containerized execution environment for each task that is separate from tasks otherwise being executed by Celery workers. Using the KubernetesPodOperator enables the following benefits:

- Task-level resource configuration: If you know how much CPU and Memory your task consumes, you can specify it

> - Execute custom Docker images on at the task level, potentially with Python packages and dependencies that would otherwise conflict with the rest of your Deployment's dependencies

Reviewer 2: looks like extra word 'on' in the clause: "Execute custom Docker images at the task level..."

- Flexibility to write task logic in a language other than Python

> - Horizontal scaling that is cost-effective, dynamic, and minimally dependent on Worker resources

Reviewer 2: Maybe add in a line about the benefit of being able to specify docker images from a private registry?

- Kubernetes-native configuration, including the ability to set volumes, secrets, and affinities in the form of a YAML file
- Coupled with Astronomer Cloud's Kubernetes-based Data Plane, run Kubernetes pods without needing to manage any underlying infrastructure

This guide provides steps for configuring and running the KubernetesPodOperator on DAGs deployed to Astronomer Cloud.

## Prerequisites

To use the KubernetesPodOperator, you need:

- An [Astronomer project](create-project).
- An Astronomer [Deployment](configure-deployment).

> - [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl).

Reviewer 2: I think kubectl is only needed if they are planning to link to a private registry, right? Maybe a note at the end of L26 mentioning that?

Reviewer 2: Addendum, looks like we have a separate "Prerequisites" section for the part about pulling images from private registries, so I think kubectl can be removed from this initial "Prereqs" section

## Set Up the KubernetesPodOperator

To use the KubernetesPodOperator in a DAG, first add the following import statements to your DAG file:

```python

> from airflow.contrib.operators.kubernetes_pod_operator import kubernetes_pod_operator

Reviewer 2: should be 
from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator

The previous link is still pre-2.0 format :/ I got the new import statement from https://registry.astronomer.io/providers/kubernetes/modules/kubernetespodoperator

# Pulls environment information from Astronomer Cloud
from airflow import configuration as conf
```

Then, instantiate the KubernetesPodOperator:

```python
from airflow.contrib.operators.kubernetes_pod_operator import kubernetes_pod_operator

# Pulls environment information from Astronomer Cloud
from airflow import configuration as conf

> ...

Reviewer 2: Are these a repeat of lines 32-37?

Author: Yeah I thought it might be good to repeat so that people see where the statements are in relation to each other / the DAG file, bu it might just be better to explain the import and the instantiation in a single code block

namespace = conf.get('kubernetes', 'NAMESPACE')
k = kubernetes_pod_operator.KubernetesPodOperator(
    namespace=namespace,
    image="<your-docker-image>",
    cmds=["bash", "-cx"],
    arguments=["echo", "10", "echo pwd"],
    labels={"<pod-Label>": "<label-name>"},
    name="<pod-name>",
    is_delete_operator_pod=True,
    in_cluster=True,
    task_id="<task-name>",
    get_logs=True)
```

For each instantiation of the KubernetesPodOperator, you must specify the following values:

- `namespace = conf.get('kubernetes', 'NAMESPACE')`: Astronomer Cloud Deployments run on their own Kubernetes namespaces within a Cluster. The KubernetesPodOperator needs to know information about this namespace. Because the namespace variable is injected into your Deployment's `airflow.cfg` file, you can programmatically import this namespace information via environment variables using this setting.

> - `image`: This is the image that the operator will use to run its defined task, commands, and arguments. By default, this value should be a publicly available image tag on Dockerhub. To pull an image from a private registry, read [Pull Images from a Private Registry](kubernetespodoperator#pull-images-from-a-private-registry).

Reviewer 2: I think I'd reword this a bit, as to me it reads like we're pointing folks towards leveraging Dockerhub. I would use language similar to what's on https://registry.astronomer.io/providers/kubernetes/modules/kubernetespodoperator, maybe something like: 
"This defaults to publicly available image tags on hub.docker.com if a fully qualified URL is not supplied. "

- `in_cluster=True`: When this value is set, your task will look inside your Cluster for a Kubernetes configuration. This ensures that the workers running in the Pod have the correct privileges within the Cluster.
- `is_delete_operator_pod=True`: This setting ensures that the Pods of completed tasks are deleted, which lowers resource usage on your Cluster.

> Once you push this code to a Deployment, your Cluster will automatically spin up and down Pods to run tasks that use the operator.

Reviewer 2: I would edit to further clarify... "to run tasks that use the KubernetesPodOperator."

  {'request_cpu': '800m',
  'request_memory': '3Gi',
  'limit_cpu': '800m',
  'limit_memory': '3Gi'}
```

To use a dictionary, specify the `resources` variable in your instantiation of the KubernetesPodOperator:

```python {11}
with dag:
    k = KubernetesPodOperator(
        namespace=namespace,
        image="hello-world",
        labels={"foo": "bar"},
        name="airflow-test-pod",
        task_id="task-one",
        in_cluster=in_cluster, # if set to true, will look in the cluster, if false, looks for file
        cluster_context='docker-for-desktop', # is ignored when in_cluster is set to True
        config_file=config_file,
        resources=compute_resources,
        is_delete_operator_pod=True,
        get_logs=True)
```

## Pull Images from a Private Registry

By default, the KubernetesPodOperator expects to pull an image that's hosted publicly on Dockerhub. If you want to pull images from a private registry, complete the setup in this topic.

### Prerequisites

To complete this setup, you need:

- A private registry that hosts your images.
- kubectl.
- An Astronomer Cloud project.
- A Deployment hosting your project.

### Step 1: Create a Kubernetes Secret

> To securely share your hosted images on Astronomer Cloud, you need to create a Kubernetes secret for accessing your private Docker registry. To do this, complete one of the following two setups:

Reviewer 2: recommend changing "share your hosted" to "leverage your private docker"

#### Option 1: Manually Create a Kubernetes Secret

1. Follow the [Kubernetes documentation](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#create-a-secret-by-providing-credentials-on-the-command-line) to create a Kubernetes secret for accessing your private registry.

2. Run the following command to confirm that the `dockerconfigjson` file was created correctly:

    ```sh

>     kubectl get secret regcred --output="jsonpath={.data.\.dockerconfigjson}" | base64 --decode

Reviewer 2: A note on this command - it's looking specifically for a secret called `regcred`. If they decide to name the secret something else when creating it in step1, this command will need to change as well. Maybe we can start of bullet point 2 with a comment like "Assuming you have named the secret you just created `regcred`"

    ```

    Copy this output.

> 3. Send the output from Step 2 and the name of the Cluster that you are configuring to your Astronomer representative. From here, your Astronomer representative will add your secret to the Cluster.

Reviewer 2: recommend changing "to your Astronomer representative" to "to Astronomer Support", and "From here, your Astronomer rep will" to "From here, Astronomer Support will"

#### Option 2: Let Astronomer Create a Kubernetes Secret

Instead of creating a Kubernetes secret yourself, you can send Astronomer the following credentials:

- Astronomer Cluster
- Docker username
- Docker password
- Docker email
- Docker server (only if other than Dockerhub)

> An Astronomer representative will use this information to directly create a Kubernetes secret and `dockerconfigjson` file in the specified Cluster.

Reviewer 2: "Astronomer Support will"

### Step 2: Run the KubernetesPodOperator

> Once Astronomer has added the Kubernetes secret to your Cluster, you will be notified and provided with the name of the secret. Typically, this secret is named `<deployment-release-name>-private-registry`.

Reviewer 2: Not really sure if we need to include the sample format of the secret name, in the 1-2 times I've done this, nobody followed that template :D

Reviewer 2: And also if we remove this, then in my opinion the example KPO task definition below should also be updated L156 with something like
image_pull_secrets=[k8s.V1LocalObjectReference('<private-registry-secret-name>')],

From here, you can use the KubernetesPodOperator with your private images by specifying `image_pull_secrets` in your operator instantiation:

```python {4}
k = kubernetes_pod_operator.KubernetesPodOperator(
    namespace=namespace,
    image="ubuntu:16.04",

>     image_pull_secrets=[k8s.V1LocalObjectReference('<deployment-release-name>-private-registry')],

Reviewer 2: Also, in order for this line to work, they'll need to add following import statement at beginning of DAG. Maybe a comment about this above, like before or after L150? 
`from kubernetes.client import models as k8s`
